Extend Request Parameters to add Body & Method

simoncaron opened the pull request:
Hey @dgtlmoon!

I found your project after it was mentioned [on the Self-Hosted podcast](https://selfhosted.show/60)! Since I'm mostly monitoring REST APIs, I felt the current HTTP GET request could benefit from a few enhancements:

- A Request Method parameter to choose the HTTP verb for the request
- A Request Body parameter which will allow to pass a JSON payload to the monitored endpoint

Since these request parameters are starting to take some place in the _General_ tab, I also propose we switch them to a new _Request_  tab which currently looks like this:

![cap2](https://user-images.githubusercontent.com/8635747/147510144-cb21841a-d55e-4c39-bfc1-8f28b85637c3.png)

Any pre-existing watch will be configured to GET by default to avoid any breaking change.

Let me know what you think, I will add some tests soon to cover these changes.

dgtlmoon commented:
Hey Simon!

Hey that's a really nice surprise! I wasnt aware of their podcast, but I had a listen and it's great! I'll be tuning in for the next one for sure!

I like your PR, it also addresses https://github.com/dgtlmoon/changedetection.io/issues/211

Two things I can think of

- I would add one thing, is that the "Body" should be disabled if the request method is `GET`, so could be an idea to add some validation to check that the `Body` field is empty if the request method is `GET`
- "A Request Body parameter which will allow to pass a JSON payload to the monitored endpoint" yeah but people are going to want to put all kinds of other data in here (like pasting some form submission thats not using a JSON content type), so I think in this case I wouldnt bother validating the JSON and just leave it to the user to trust they put in the right data (and make it more flexible by allowing any other content type)

And yes - it needs tests or it doesnt get merged 8)

Thanks for the PR!

dgtlmoon commented:
Once its merged, I will also move the fetch-method to your `requests` tab

![image](https://user-images.githubusercontent.com/275001/147552972-fade06b1-22bc-4c72-a999-f501484e1bf4.png)

dgtlmoon commented:
By the way, thanks for coping with my janky test harness :) feel free to split your test off into its own `test_fetch_method_verb.py` or something, can be cleaner/simpler than one of the existing (and very long) tests

simoncaron commented:
I was on the fence for the JSON validation too... I've removed it and added a validation step to ensure the body field is empty on GET.

I also added a few tests for these new fields and conditions.

dgtlmoon commented on changedetectionio/fetch_site_status.py:
```diff
@@ -80,6 +80,8 @@ def run(self, uuid):
         else:
             timeout = self.datastore.data['settings']['requests']['timeout']
             url = self.datastore.get_val(uuid, 'url')
+            request_body = self.datastore.get_val(uuid, 'body')
```
sorry to be annoying, any chance this can `request_body`

dgtlmoon commented on changedetectionio/fetch_site_status.py:
```diff
@@ -80,6 +80,8 @@ def run(self, uuid):
             timeout = self.datastore.data['settings']['requests']['timeout']
             url = self.datastore.get_val(uuid, 'url')
+            request_body = self.datastore.get_val(uuid, 'body')
+            request_method = self.datastore.get_val(uuid, 'method')
```
and sorry to be annoying, any chance this can `request_method`

dgtlmoon commented:
Ok i just realised that I already have `headers` so asking to change to `request_body` is kind of mute, i'm already using a not so specific field name

leikoilja commented:
This is amazing, thank you!

Successfully used it to perform `POST` requests with form data to some old registration page that I need to monitor.
For anyone coming here for that, here are some useful tips
![image](https://user-images.githubusercontent.com/10655107/163115664-ec87eecc-6860-43ad-8abc-a68d0e782763.png)
![image](https://user-images.githubusercontent.com/10655107/163115820-4e3ffc63-c077-4091-b55b-f1a40e1a9b54.png)

dgtlmoon commented:
@leikoilja nice documentation is just as important as nice code :D can you copy your text above there to https://github.com/dgtlmoon/changedetection.io/wiki/Using-POST-PUT-PATCH-DELETE-requests  (just made that page) , that would make me very happy :)
